test/metabase/test/data/mbql_query_impl.clj

```diff
@@ -8,7 +8,18 @@
 
 (set! *warn-on-reflection* true)
 
-(def ^:dynamic *id-fn-symb*              'metabase.test.data/id)
+#_:clj-kondo/ignore
+(defn druid-id-fn
+  "I have a strong feeling this should be handled differently! Let's discuss that during review!"
+  [& args]
+  (if (and (> (count args) 1)
+           (= :druid-jdbc @(requiring-resolve 'metabase.driver/*driver*))
+           (= "timestamp" (name (last args))))
+    (apply @(requiring-resolve 'metabase.test.data/id) (conj (vec (butlast args)) :__time))
+    (apply @(requiring-resolve 'metabase.test.data/id) args)))
+
+;; TODO: druid-id-fn is just temporary until I figure out proper solution for rebinding that symb.
+(def ^:dynamic *id-fn-symb*              `druid-id-fn #_'metabase.test.data/id)
```

Comment:
I'd like to rebind this for eg. mbql-query execution in timeseries tests and make that druid-id-fn local to that ns, but I was unable to find a way of doing that. Hence I've resorted to this adhoc implementation. I'd welcome some pointers on how to achieve that. Thanks.

Or I could remove the __time to timestamp column name conversion from Druid driver.

Reply:
In the timeseries tests can't just just create your own `run-mbql-query` macro like 

```clj
(defmacro run-mbql-query [& args]
  `(binding [mbql-query-impl/*id-fn-symb* <whatever>]
     (mt/run-mbql-query ~@args)))
```

Reply:
I was unable to make the snippet work as intended. See the follwing:

```
;; definition
(defmacro run-mbql-query* [& args]
  `(binding [metabase.test.data.mbql-query-impl/*id-fn-symb* ~druid-id-fn]
     (mt/run-mbql-query ~@args)))
     
;; Let's expand simple query
(clojure.walk/macroexpand-all '(run-mbql-query* checkins {:fields [$timestamp] :limit 1}))

;; expansion result
(let*
 []
 (clojure.core/push-thread-bindings
  (clojure.core/hash-map
   #'metabase.test.data.mbql-query-impl/*id-fn-symb*
   #function[metabase.timeseries-query-processor-test/druid-id-fn]))
 (try
  (metabase.test.data/run-mbql-query*
   {:database (metabase.test.data/id),
    :type :query,
    :query
    {:fields [[:field (metabase.test.data/id :checkins :timestamp) nil]],
     :limit 1,
     :source-table (metabase.test.data/id :checkins)}})
  (finally (clojure.core/pop-thread-bindings))))
```

It seems that binding is expanded into `push-thread-bindings` that goes into effect during execution, but `metabase.test.data.mbql-query-impl/*id-fn-symb*` is used during macro expansion I believe.

I've also tried to approach it using lower level `push-thread-bindings` directly instead of `binding`, but without success. Here unfortunately I have no idea how to make it work without modifications to `mbql-query-impl` namespace. Or is my reasoning erroneous here?

Reply:
Oops, sorry I forgot about that. Those bindings have to be bound during macro expansion, which means you have to have the macro that binds them return the fully-expanded form rather than another macro that gets expanded latter.

So you have a few options here

## Option 1

One is to just have your macro call `macroexpand` itself like this

```clj
(defn druid-id [& args]
  (into [:druid-id] args))
  
(defmacro run-mbql-query {:style/indent :defn} [& args]
  (binding [mbql-query-impl/*id-fn-symb* `druid-id]
    (macroexpand `(mt/mbql-query ~@args))))

(macroexpand
 (run-mbql-query venues
   {:fields [$id]}))
;; => 
{:database [:druid-id]
 :type     :query
 :query     {:fields       [[:field [:druid-id :venues :id] nil]]
             :source-table [:druid-id :venues]}}
```

## Option 2 

Another option is just to have your macro call the same underlying functions that `mbql-query` is implemented with. That's how we implement the `metabase.lib.test-util.macros/mbql-query` macro which binds `*id-fn-symb*; it looks like this:

```clj
(defmacro mbql-query
  "MLv2 version of [[metabase.test/mbql-query]] that uses the [[metabase.lib.test-metadata]] rather than the application
  database."
  {:style/indent :defn}
  ([table-name]
   `(mbql-query ~table-name {}))

  ([table-name inner-query]
   {:pre [(map? inner-query)]}
   (do-with-bindings
    #(as-> inner-query <>
       (mbql-query-impl/parse-tokens table-name <>)
       (mbql-query-impl/maybe-add-source-table <> table-name)
       (mbql-query-impl/wrap-inner-query <>)))))
```

## Option 3

I think the last option is admitting that using dynamic variables to control macroexpansion behavior is dumb; we can add another internal macro called something like `mbql-query-with-options` that takes the usual `mbql-query`/`run-mbql-query` args **plus** and options map with stuff like the `id-fn-symb`. `mbql-query` can just call `mbql-query-with-options` with the default options, and `metabase.lib.test-util.macros/mbql-query` and your version of `run-mbql-query` can call it with their respective custom options.

Reply:
I think Option 3 is the right way to do things, but this PR is already big enough without deciding to refactor test utils on top of it, so if I were you I would probably just go with Option 1 since it's the easiest and maybe we can refactor this stuff in the future.

Reply:
Honestly tho it's just test tooling so if it works then I don't think we need to sweat the details about making everything perfect